README.md
# Autoplaylists for Google Music™

![Repominder](https://img.shields.io/badge/dynamic/json.svg?label=release&query=%24.status&maxAge=43200&uri=https%3A%2F%2Fwww.repominder.com%2Fbadge%2FeyJ1c2VyX2lkIjogMiwgInJlcG9faWQiOiAxOH0%3D%2F&link=https%3A%2F%2Fwww.repominder.com%2F "Repominder")

This is a Chrome extension to add user-defined autoplaylists in Google Music (iTunes calls these "Smart Playlists").

![autoplaylist screenshot](autoplaylists_screenshot.png?raw=true)

**This extension is not supported nor endorsed by Google.**

## Installation and support
Everything relevant to users can be found at https://autoplaylists.simon.codes.

The rest of this page is intended for developers.

## Development workflow
* run `bundle install`
* run `npm install`
* run `./watch.sh` somewhere and leave it running -- this will build whenever files change
* go to chrome://extensions/
* click "developer mode"
* click "Load unpacked extension" and provide the repo/src directory
* after updating any javascript, hit "reload" in chrome://extensions/
* you can find the logs in different places depending on the code that's running:
    * content script: music.google.com console
    * background script: click the "background page" link by the "Inspect views" section on chrome://extensions/
    * manager/playlist: on their pages' consoles

---
Google Music is a trademark of Google Inc. Use of this trademark is subject to Google Permissions.


CHANGELOG.md
## 5.10.10 (May 5, 2020)

Misc:

  - add a confirmation check when deleting autoplaylists

## 5.10.9 (March 23, 2020)

Misc:

  - further clarify the multi-user error message

## 5.10.8 (October 28, 2019)

Misc:

  - suggest turning on sync in multi-user error message

## 5.10.7 (August 18, 2019)

Bugfixes:

  - fix initialization of large libraries from Google's local store

## 5.10.6 (November 20, 2018)

Misc:

  - reduce required permissions

## 5.10.5 (November 20, 2018)

Misc:

  - various dependency upgrades; no user-facing changes

## 5.10.4 (August 5, 2018)

Misc:

  - turn on change batching for everyone to improve sync success rates
  - improve the presentation of the backoff status

## 5.10.3 (July 16, 2018)

Bugfixes:

  - prevent error responses from stopping the sync queue

## 5.10.2 (July 15, 2018)

Bugfixes:

  - add an additional validation step to prevent unauthed tokens reported as valid

Misc:

  - update to the latest android protocol

## 5.10.1 (July 14, 2018)

Misc:

  - improve the timing of sync page polling
  - remove some unneeded error reporting

## 5.10.0 (July 14, 2018)

Features:

  - the syncing page now surfaces backoff status

Bugfixes:

  - the syncing page properly handles timeouts

Misc:

  - mutation response codes are logged and reported, which should help with sync debugging

## 5.9.0 (June 17, 2018)

Features:

  - a syncing page has been added, which contains information on past syncs and buttons to manually trigger them

Bugfixes:

  - properly reset random orderings on tab refreshes

## 5.8.1 (June 2, 2018)

Misc:

  - playlists are now always synced when hitting "save", even if their definition is unchanged

## 5.8.0 (May 28, 2018)

Misc:

  - playlists are now sorted when presented for playlist equal/not-equal conditions

## 5.7.0 (March 11, 2018)

Misc:

  - overhaul the UI

## 5.6.1 (January 14, 2018)

Misc:

  - temporarily link the user survey under the version blurb

## 5.6.0 (January 13, 2018)

Features:

  - display autoplaylists in alphabetic order to match Google's presentation order

## 5.5.5 (December 19, 2017)

Misc:

  - fix packaging of jekyll html for real this time

## 5.5.4 (December 19, 2017)

Misc:

  - fix packaging of jekyll html

## 5.5.3 (December 19, 2017)

Misc:

  - switch to jekyll behind the scenes for html templating; should not have any noticeable affects

## 5.5.2 (December 16, 2017)

Bugfixes:

  - increase the default sync interval to 20m; add a note about low intervals being risky

## 5.5.1 (December 12, 2017)

Bugfixes:

  - improve reliability of tab detection

## 5.5.0 (November 13, 2017)

Features:

  - add a link that discards changes when adding/creating an autoplaylist (Dasonk, [#175](https://github.com/simon-weber/Autoplaylists-for-Google-Music/pull/175))

## 5.4.6 (November 11, 2017)

Misc:

  - prevent trial expiration notification from showing to full users after downgrading

## 5.4.5 (October 7, 2017)

Misc:

  - error reporting tweaks

## 5.4.4 (October 1, 2017)

Bugfixes:

  - adapt to Google api change to fix remote playlists not being deleted along with their autoplaylist

## 5.4.3 (October 1, 2017)

Misc:

  - add a one week free trial of the full version

## 5.4.2 (August 7, 2017)

Misc:

  - disallow automatic reloads within the first 10m of starting (to prevent reload loops)

## 5.4.1 (August 7, 2017)

Bugfixes:

  - fix license caching bug that prevented re-querying the license api in some cases

Misc:

  - show a notification to users who installed before the mailing list existed so they're aware of it

## 5.4.0 (August 5, 2017)

Features:

  - add two new boolean conditions: "has album art?" and "is matched?"

Bugfixes:

  - fix playlist linking conditions that exclude tracks (eg 'playlist is not...' and 'playlist title doesn't match...')
  - fix an endless loop involving playlists that indirectly link to themselves

## 5.3.6 (July 17, 2017)

Misc:

  - switch announcements link to mailchimp archive

## 5.3.5 (July 15, 2017)

Misc:

  - add a mailchimp prompt to the welcome page

## 5.3.4 (July 13, 2017)

Misc:

  - improve reliability over long periods by automatically reloading after deauths

## 5.3.3 (July 12, 2017)

Bugfixes:

  - fix periodic sync stops caused by closed tabs

Misc:

  - automatically connect to a GM tab on load (previously a refresh was required)

## 5.3.2 (July 11, 2017)

Bugfixes:

  - fix playlists using a "playlist title" condition that matched non-autoplaylists

## 5.3.1 (June 28, 2017)

Misc:

  - switch to a different chrome api for communicating with the Google Music tab (which required adding the tabs permission)

## 5.3.0 (June 17, 2017)

Features:

  - show test results with limit applied by default; add a button to show results without the limit

## 5.2.2 (June 16, 2017)

Bugfixes:

  - reduce the likelyhood of hitting sync storage write quotas when deleting playlists

## 5.2.1 (May 13, 2017)

Misc:

  - add a warning that uninstalling will delete definitions
  - error reporting tweaks

## 5.2.0 (March 19, 2017)

Features:

  - add a condition for linking playlists via their title
  - add a button to manually trigger a sync

Misc:

  - improve error reporting

## 5.1.6 (March 5, 2017)

Misc:

  - add an experimental option to send multiple smaller requests instead of one large one. See https://github.com/simon-weber/Autoplaylists-for-Google-Music/issues/128#issuecomment-283096925 for details.
  - rename "track" and "disc" to "track number" and "disc number"
  - update description format

## 5.1.5 (February 26, 2017)

Bugfixes:

  - prevent overloading Google's servers by pausing syncing for 15 minutes when they return an error

## 5.1.4 (January 28, 2017)

Bugfixes:

  - fix incorrect results for some autoplaylists that include multiple other autoplaylists

## 5.1.3 (January 2, 2017)

Bugfixes:

  - fix sync problems for users with a system clock ahead of Google's

## 5.1.2 (January 1, 2017)

Bugfixes:


js/track.js
'use strict';

const Lf = require('lovefield');

function f(fieldItems) {
  return {
    protoNum: fieldItems[0],
    name: fieldItems[1],
    type: fieldItems[2],
    label: fieldItems[3] || fieldItems[1],
  };
}

exports.fields = [
  f([0, 'id', Lf.Type.STRING]),
  f([1, 'title', Lf.Type.STRING]),
  f([3, 'artist', Lf.Type.STRING]),
  f([4, 'album', Lf.Type.STRING]),
  f([5, 'albumArtist', Lf.Type.STRING, 'album artist']),
  f([10, 'composer', Lf.Type.STRING]),
  f([11, 'genre', Lf.Type.STRING]),
  f([13, 'durationMillis', Lf.Type.INTEGER]),
  f([14, 'track', Lf.Type.INTEGER]),
  f([15, 'totalTracks', Lf.Type.INTEGER]),
  f([16, 'disc', Lf.Type.INTEGER]),
  f([17, 'totalDiscs', Lf.Type.INTEGER]),
  f([18, 'year', Lf.Type.INTEGER]),
  f([22, 'playCount', Lf.Type.INTEGER, 'play count']),
  f([23, 'rating', Lf.Type.INTEGER]),
  f([24, 'creationDate', Lf.Type.INTEGER]),
  f([25, 'lastPlayed', Lf.Type.INTEGER]),
  f([29, 'type', Lf.Type.INTEGER]),
  f([30, 'comment', Lf.Type.STRING]),
  f([34, 'bitrate', Lf.Type.INTEGER]),
];

const lToB = {};
lToB[Lf.Type.STRING] = 'string';
lToB[Lf.Type.INTEGER] = 'numeric';

exports.lfToBusinessTypes = lToB;

exports.fromJsproto = function fromJsproto(jsproto) {
  const track = {};
  exports.fields.forEach(field => {
    track[field.name] = jsproto[field.protoNum];
  });

  return track;
};
